### PR TITLE
add tg-mxfp4-moe-test

### PR DESCRIPTION
**Pull request by IwakuraRein**
# Essential Elements of an Effective PR Description Checklist

- Add unit tests for mxfp4 x mxfp8 and mxfp4 x bf16 trtllm-gen moe.

- [x] The test plan, such as providing test command.
- [x] The test results, such as pasting the results comparison before and after, or e2e results
- [x] (Optional) The necessary documentation update, such as updating `supported_models.md` and `examples` for a new model.

### CODE REVIEW

File: tests/kernels/moe/test_mxfp4_moe.py [hunk 1]
```diff
@@ -4,9 +4,13 @@
 import importlib
 import importlib.metadata
 from dataclasses import dataclass
+from typing import Optional
 
 import pytest
 import torch
+from flashinfer import (fp4_quantize, mxfp8_quantize, next_positive_power_of_2,
+                        reorder_rows_for_gated_act_gemm, shuffle_matrix_a,
+                        shuffle_matrix_sf_a, trtllm_fp4_block_scale_moe)
```
Comment (mgoin) [inline tests/kernels/moe/test_mxfp4_moe.py]: It looks like there are potentially amd tests in this file, so could we wrap the flashinfer import in a try or just include it in the functions that need it?

File: tests/kernels/moe/test_mxfp4_moe.py [hunk 2]
```diff
@@ -54,4 +58,410 @@ def test_mxfp4_loading_and_execution_moe(vllm_runner, model_case: ModelCase):
...
+            f"(threshold: {1-percent:.4f})")
+
+
+@pytest.mark.parametrize("topk", [1, 4])
+@pytest.mark.parametrize("num_experts", [32, 128])
+@pytest.mark.parametrize("num_tokens", [1, 128, 1024])
+@pytest.mark.parametrize("intermediate_size,hidden_size", [(3072, 3072)])
+@pytest.mark.parametrize("alpha,limit", [(1.0, None), (1.702, 7.0)])
+@pytest.mark.parametrize("act_type", ['mxfp8', 'bf16'])
+def test_trtllm_gen_mxfp4_fused_moe(
```
Comment (mgoin) [inline tests/kernels/moe/test_mxfp4_moe.py:362]: What compute capability is required for this moe? I assumed the trtllm one is only SM100, so we should have a skip if not current_platform.is_device_capability(100)

File: tests/kernels/moe/test_mxfp4_moe.py [hunk 3]
```diff
@@ -54,4 +58,410 @@ def test_mxfp4_loading_and_execution_moe(vllm_runner, model_case: ModelCase):
...
+if __name__ == "__main__":
+    torch.set_printoptions(threshold=1000, sci_mode=False, precision=3)
+    test_trtllm_gen_mxfp4_fused_moe(topk=4,
+                                    num_experts=32,
+                                    num_tokens=128,
+                                    intermediate_size=4096,
+                                    hidden_size=4096,
+                                    act_type='mxfp8',
+                                    alpha=1.0,
+                                    limit=None)
```
Comment (mgoin) [inline tests/kernels/moe/test_mxfp4_moe.py]: Remove this to just leave the test cases

**Comment by mgoin** [comment]
Thanks for the kernel tests! LGTM just a few comments

**Review by zyongye — APPROVED** [review]
LGTM

File: tests/kernels/moe/test_mxfp4_moe.py [hunk 4]
```diff
@@ -13,6 +14,16 @@
...
 
+device_props = torch.cuda.get_device_properties(torch.cuda.current_device())
+TRTLLM_GEN_MXFP4_AVAILABLE = torch.cuda.is_available(
+) and device_props.major == 10 and device_props.minor == 0
+
+if TRTLLM_GEN_MXFP4_AVAILABLE:
+    from flashinfer import (fp4_quantize, mxfp8_quantize,
+                            next_positive_power_of_2,
+                            reorder_rows_for_gated_act_gemm, shuffle_matrix_a,
+                            shuffle_matrix_sf_a, trtllm_fp4_block_scale_moe)
```
Comment (mgoin) [inline tests/kernels/moe/test_mxfp4_moe.py]: You can use current_platform for this i.e.
```python
from vllm.platforms import current_platform

TRTLLM_GEN_MXFP4_AVAILABLE = current_platform.is_cuda() and current_platform.is_device_capability(100)
```

Also please add this file to https://github.com/vllm-project/vllm/blob/1891a265d316217f9c1e552cf7c380ef5bf1eec1/.buildkite/test-pipeline.yaml#L648 in order to add it to the B200 runner we have

**Review by mgoin — APPROVED** [review]

File: tests/kernels/moe/test_mxfp4_moe.py [hunk 5]
```diff
@@ -54,4 +66,410 @@ def test_mxfp4_loading_and_execution_moe(vllm_runner, model_case: ModelCase):
...
+    assert a.shape == b.shape, f"Shape mismatch: {a.shape} vs {b.shape}"
+
+    left = torch.abs(a - b)
+    right = atol + rtol * torch.abs(b)
+    count = torch.sum(left > right)
+    mismatch_percent = count / a.numel()
+    if mismatch_percent > 1 - percent:
+        raise Exception(
+            f"Mismatch percentage is {mismatch_percent:.4f} for rtol {rtol} "
+            f"(threshold: {1-percent:.4f})")
```
Comment (yewentao256) [inline tests/kernels/moe/test_mxfp4_moe.py:349]: Is there any reason we don't use torch_assert_close?

**Review by yewentao256 — APPROVED** [review]
LGTM, thanks for the work!

File: tests/kernels/moe/test_mxfp4_moe.py (hunk 5 above)
Comment (IwakuraRein) [reply tests/kernels/moe/test_mxfp4_moe.py:349]: @yewentao256 Hi. Currently the result from fp4 moe is not exactly the same as reference result even if I quantize the reference weight so we used a loose accuracy check instead of `assertall_close`. This is probably due to the difference in the sequence of calculations between fp4 moe and reference moe. We are still investigating the issue.

Comment (yewentao256) [reply tests/kernels/moe/test_mxfp4_moe.py:349]: Make sense, please let me know if any further investigation results available.
